saga/seata-saga-engine/src/main/java/io/seata/saga/engine/expression/impl/DefaultExpressionResolver.java

```diff
@@ -0,0 +1,138 @@
+/*
+ *  Copyright 1999-2019 Seata.io Group.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package io.seata.saga.engine.expression.impl;
+
+import io.seata.saga.engine.expression.Expression;
+import io.seata.saga.engine.expression.ExpressionFactory;
+import io.seata.saga.engine.expression.ExpressionFactoryManager;
+import io.seata.saga.engine.expression.ExpressionResolver;
+import io.seata.saga.engine.expression.exception.ExceptionMatchExpression;
+import io.seata.saga.engine.expression.spel.SpringELExpression;
+import io.seata.saga.proctrl.ProcessContext;
+import io.seata.saga.statelang.domain.DomainConstants;
+
+/**
+ * Default {@link ExpressionResolver} implementation
+ *
+ * @author ptyin
+ */
+public class DefaultExpressionResolver implements ExpressionResolver {
+
+    protected static class ExpressionStruct {
+        int typeStart;
+
+        int typeEnd;
+
+        int end;
+
+        String type;
+
+        String content;
+    }
+
+    private ExpressionFactoryManager expressionFactoryManager;
+
+    private String rootObjectName;
+
+    @Override
+    public Expression getExpression(String expressionStr) {
+        ExpressionStruct struct = parseExpressionStruct(expressionStr);
+
+        ExpressionFactory expressionFactory = expressionFactoryManager.getExpressionFactory(struct.type);
```

Comment:
we need to compatible history expressions without type. so if expression type is null, default expression type?

Reply:
It works fine with passing `null`